_posts/2017-09-04-Secure-your-architecture-part1.md

```diff
@@ -0,0 +1,427 @@
+The OAuth2 starter includes the JWT dependency of Spring Security.
+Secondly, we make a class to decode our incoming JWT.
+
+{% highlight java %}
+
+public class JwtAccessTokenConvertDecoder extends JwtAccessTokenConverter {
+
+    private JsonParser objectMapper = JsonParserFactory.create();
+
+    protected Map<String, Object> decode(String token) {
+        try {
+            Jwt jwt = JwtHelper.decode(token);
+            String content = jwt.getClaims();
+            Map<String, Object> map = objectMapper.parseMap(content);
+            if (map.containsKey(EXP) && map.get(EXP) instanceof Integer) {
+                Integer intValue = (Integer) map.get(EXP);
+                map.put(EXP, Long.valueOf(intValue));
+            }
+            return map;
+        }
+        catch (Exception e) {
+            throw new InvalidTokenException("Cannot convert access token to JSON", e);
+        }
+    }
+}
+
+{% endhighlight %}
+
+And add these as beans to get it working. 
+
+{% highlight java %}
+@Configuration
+@EnableResourceServer
+public class ResourceServiceConfiguration extends ResourceServerConfigurerAdapter {
+
+    @Bean
+    public JwtAccessTokenConvertDecoder accessTokenConverter() {
+        return new JwtAccessTokenConvertDecoder();
+    }
+
+    @Bean
+    public TokenStore tokenStore() {
+        return new JwtTokenStore(accessTokenConverter());
+    }
+
+    @Bean
+    @Primary
+    public DefaultTokenServices tokenServices() {
+        DefaultTokenServices defaultTokenServices = new DefaultTokenServices();
+        defaultTokenServices.setTokenStore(tokenStore());
+        return defaultTokenServices;
+  }
+}
+
+{% endhighlight %}
+
+### Common mistakes with keys
+So the UAA signs the token with a private key and verifies it with a public key. 
+The problem that occurs is that every microservice would need the public key to connect with the UAA server for verification on every request.
+As you can see, we don't want every microservice to be dependant of the UAA server to be available at startup, in testing or CI. 
+The solution to this is to disable exposure of your microservices to the outer network and handle only incoming traffic via the reverse proxy/load balancer (Zuul).
+Zuul will verify the token as a trustworthy client of the UAA server and will propagate the token to the downstreamed services.
+The microservice will handle the token as trustworthy since the token is verified.
+
+
+### Securing API endpoints
+At last we're going to secure our resources.
+Spring Security gives us a variety of tools to secure at class/method level.
+Let me point you out the most used one.  <br />
+To enable method security add `@EnableGlobalMethodSecurity(prePostEnabled = true)` to your configuration. 
+
+#### Authority 
+For the authorization, Spring Security provides us authorities, extracted from the JWT, the authorities are placed inside a principal.
```

Comment:
Zinnen een beetje opkuisen:
For the authorization, Spring Security provides us _with_ authorities, extracted from the JWT.
These are placed inside a principal, which will be used throughout the existing security context of your application.
You can then reference them using Spring Expression Language (SpEL) to secure your methods.
There are plenty of options you can use for method security, but we'll highlight the most common ones.
You can find a complete list in the [Spring documentation](https://docs.spring.io/spring-security/site/docs/3.0.x/reference/el-access.html).